source/mir/stat/distribution/binomial.d

```diff
@@ -0,0 +1,213 @@
+/++
+This module contains algorithms for the binomial probability distribution.
+
+License: $(HTTP www.apache.org/licenses/LICENSE-2.0, Apache-2.0)
+
+Authors: John Michael Hall
+
+Copyright: 2022 Mir Stat Authors.
+
++/
+
+module mir.stat.distribution.binomial;
+
+import mir.bignum.fp: Fp;
+import mir.internal.utility: isFloatingPoint;
+
+/++
+Computes the binomial probability mass function (PMF).
+
+This function can control the type of the function output through the template
+parameter `T`. By default, `T` is set equal to `double`, but other floating
+point types or extended precision floating point types (e.g. `Fp!128`) can be
+used. For large values of `n`, `Fp!128` is recommended.
+
+Params:
+    k = value to evaluate PMF (e.g. number of "heads")
+    n = number of trials
+    p = `true` probability
+
+See_also:
+    $(LINK2 https://en.wikipedia.org/wiki/Binomial_distribution, binomial probability distribution)
++/
+@safe pure nothrow @nogc
+T binomialPMF(T = double, U)(const uint k, const uint n, const U p)
```

Comment:
@9il If changing to `binomialPMF(T = U, U)`,  then you always have to specify the return type (i.e. `T`). Otherwise, you will get errors about `U` being undefined. 

```d
T foo(T = U, U)(U x) {
 	return cast(T) x;   
}

void main() {
    double a = 3.0;
    auto x = foo!int(a);
    auto y = foo(a);
}
```

If you re-organize it so that `binomialPMF(U = double, T = U)` then it makes it harder to use since specifying the return type comes second.

Reply:
@9i I get errors about incompatible types when doing `1 - p` and in the comparisons. The following is a reduced example of the errors. I tried to throw some casts in there without success (I can't cast an `int` to an `Fp!size`, but I can cast a `double`, but then I can't do subtraction between two `Fp`s). When I do the same for `bar`, I get an error that `Fp!size` doesn't implement `opCmp`. 

```d
import mir.bignum.fp: Fp;

Fp!128 foo(Fp!128 x) {
     return 1 - x;   
}

Fp!128 bar(Fp!128 x)
    in (x > 0)
{
    return x;   
}

void main() {
    auto a = Fp!128(0.75);
    auto x = foo(a);
    auto y = bar(a);
}
``